src/inmanta/ast/statements/assign.py

```diff
@@ -204,9 +249,32 @@ def _add_to_dataflow_graph(self, graph: typing.Optional[DataflowGraph]) -> None:
         node: dataflow.AttributeNodeReference = self.instance.get_dataflow_node(graph).get_attribute(self.attribute_name)
         node.assign(self.value.get_dataflow_node(graph), self, graph)
 
+    def emit_progression_promises(
+        self, resolver: Resolver, queue: QueueScheduler, *, in_scope: FrozenSet[str], root: bool = True
+    ) -> Sequence[ConditionalPromiseABC]:
+        if root:
+            return []
+        # TODO: handle explicit namespace references, are they always in scope (think not)? If so, add to parent docstring.
+        #       If not, how to deal with them?
+        # TODO: this check is a temporary hack, clean it up!
+        if "::" in self.instance.name or self.instance.name.split(".", 1)[0] not in in_scope:
+            return []
+
+        promise: ConditionalPromise = ConditionalPromise(provider=self)
+        hook: VariableReferenceHook = VariableReferenceHook(
```

Comment:
This has the exact same structure as the first stage of execution.

Would it be more interesting to 'convert' promises when picking them? instead of re-emitting? 
It would also eliminate the problem of in-accurate tracking because of a race between removal of a promise through picking and re-emit of the non-conditional equivalent?